### PR TITLE
Fix issues with types and comparisons

### PR DESCRIPTION
**Pull request by adamretter**
Closes https://github.com/eXist-db/exist/issues/1176

Cleans up value and general comparisons to bring them inline with the XQuery spec. 

Some highlights:

1. eXist-db's Type Hierarchy had a bit of a nasty hack around numeric types because it was unable to represent union types. Previously eXist-db modelled `xs:decimal`, `xs:float`, and `xs:double`, as sub-types of the union type `xs:numeric` which was incorrect. The XDM specifies that they are sub-types of `xs:anyAtomicType`, and that `xs:numeric` is a union type whose members are the set of `xs:decimal`, `xs:float`, and `xs:double`. This PR adds support for modelling Union Types from the XDM and corrects the type hierarchy in eXist-db. See - https://www.w3.org/TR/xpath-datamodel-31/#types-hierarchy

2. When performing Value and General comparisons, eXist-db's type coercion was incorrect. The XQuery spec specifies the following algorithms: [Value Comparisons](https://www.w3.org/TR/xquery-31/#id-value-comparisons), and [General Comparisons](https://www.w3.org/TR/xquery-31/#id-general-comparisons). This PR replaces the existing type coercion in eXist-db with a new version which is inline with the XQuery specification.

3. Comparisons between the values backing numbers in eXist-db were always down-cast to a Java `double` value. This meant a loss of precision when comparing larger numbers. This PR replaces that with an up-cast and lambda pass method; This accommodates the smallest common Java type, thus ensuring no truncation/rounding of numeric data.

### CODE REVIEW

**Comment by joewiz** [comment]
@adamretter Wow, thanks for diving into that challenge. I'm curious, does XQTS give you any indication about improvements (or even possibly regressions, I suppose) that result from this PR?

**Comment by adamretter** [comment]
> does XQTS give you any indication about improvements (or even possibly regressions, I suppose) that result from this PR?

@joewiz actually we get a small regression in XQTS, but I think that is as a product of other bugs in eXist-db, whereby we used to pass tests that we shouldn't of.  Before this PR we had `78.92%`, after this PR we have `78.49%`. So we will have to chase those other bugs in future PRs ;-)

File: exist-core/src/main/java/org/exist/xquery/GeneralComparison.java [hunk 1]
```diff
@@ -1040,6 +1041,48 @@ private AtomicValue convertForCompareAtomic(final AtomicValue value, final int t
...
+            if ((Type.subTypeOf(thisType, Type.DECIMAL) || thisType == Type.FLOAT)
+                    && (Type.subTypeOf(otherType, Type.DECIMAL) || otherType == Type.FLOAT)) {
+                return value.convertTo(Type.FLOAT);
+            }
+
+            /*
+             *  c. If each operand is an instance of one of the types xs:decimal, xs:float, or xs:double,
+             *     then both operands are cast to type xs:double.
+             */
+            if ((Type.subTypeOf(thisType, Type.DECIMAL) || thisType == Type.FLOAT || thisType == Type.DOUBLE)
```
Comment (line-o) [inline exist-core/src/main/java/org/exist/xquery/GeneralComparison.java:1073]: Note to self: Test for behaviour with decimal subtypes and also subtypes of xs:integer

**Comment by line-o** [comment]
retest #3330

**Comment by wolfgangmm** [comment]
With this PR applied, I see the following NPE in the logs:

```
Caused by: java.lang.NullPointerException
	at org.exist.xquery.DynamicTypeCheck.check(DynamicTypeCheck.java:84) ~[exist-core-5.3.0-SNAPSHOT.jar:5.3.0-SNAPSHOT]
	at org.exist.xquery.DynamicTypeCheck.eval(DynamicTypeCheck.java:67) ~[exist-core-5.3.0-SNAPSHOT.jar:5.3.0-SNAPSHOT]
	at org.exist.xquery.FunctionCall.eval(FunctionCall.java:205) ~[exist-core-5.3.0-SNAPSHOT.jar:5.3.0-SNAPSHOT]
```

**Comment by adamretter** [comment]
@wolfgangmm that's not great... But you need to tell me how to reproduce it for me to help!

**Comment by wolfgangmm** [comment]
@adamretter sure, this was just a notice. Debugging takes a while. The error does not pop up immediately, otherwise it would be simple.

**Comment by dizzzz** [comment]
image

almost !!!!

**Comment by dizzzz** [comment]
@adamretter the Java13 issue seems to be real

**Comment by wolfgangmm** [comment]
The NPE I reported does not seem related to this PR. The issue also appears on older eXist versions, but only under very particular circumstances. I'll submit a separate fix.

**Comment by joewiz** [comment]
@adamretter This PR doesn't address https://github.com/eXist-db/exist/issues/1715, does it?

Also, does anyone have insights into the failing Travis openjdk13 test?

**Comment by adamretter** [comment]
@joewiz 

> This PR doesn't address #1715, does it?

No.

> Also, does anyone have insights into the failing Travis openjdk13 test?

Yes, I have a reproducible test case here that I am working with.

File: exist-core/src/main/java/org/exist/xquery/value/DoubleValue.java [hunk 2]
```diff
@@ -39,7 +41,7 @@
     public static final DoubleValue NEGATIVE_INFINITY = new DoubleValue(Double.NEGATIVE_INFINITY);
     public static final DoubleValue NaN = new DoubleValue(Double.NaN);
 
-    private final double value;
+    final double value;
```
Comment (dizzzz) [inline exist-core/src/main/java/org/exist/xquery/value/DoubleValue.java:44]: package protected on purpose?

File: exist-core/src/main/java/org/exist/xquery/value/Type.java [hunk 3]
```diff
@@ -105,95 +105,136 @@
...
 
-        //THIS type system is broken - some of the below should be sub-types of ANY_SIMPLE_TYPE
-        //and some should not!
+        // ITEM sub-types
         defineSubType(ITEM, ATOMIC);
-        defineSubType(ATOMIC, STRING);
-        defineSubType(ATOMIC, BOOLEAN);
-        defineSubType(ATOMIC, QNAME);
+        defineSubType(ITEM, FUNCTION_REFERENCE);
+        //defineSubType(ITEM, NODE);                // TODO(AR) this appears in the XDM 3.1, but uncommenting this breaks a lot of stuff in eXist-db
```
Comment (line-o) [inline exist-core/src/main/java/org/exist/xquery/value/Type.java:135]: Setting NODE to a different value (I tried 7, the first available number not yet taken) allows this definition.
https://github.com/eXist-db/exist/pull/3363/files#diff-14a14d3b3adcc1ebc87fe76d1e77c862R43
At least all Xquery3 Tests pass.

File: exist-core/src/main/java/org/exist/xquery/value/DoubleValue.java (hunk 2 above)
Comment (adamretter) [reply exist-core/src/main/java/org/exist/xquery/value/DoubleValue.java:44]: I think it's needed in `#createComparisonWith(NumericValue)`, did you see otherwise?

File: exist-core/src/main/java/org/exist/xquery/value/Type.java (hunk 3 above)
Comment (adamretter) [reply exist-core/src/main/java/org/exist/xquery/value/Type.java:135]: Hmmm... okay that's interesting. I would have thought that changing its value should have no effect! Thanks Juri, I will take a look at this...

Comment (line-o) [reply exist-core/src/main/java/org/exist/xquery/value/Type.java:135]: It was the only negative value, therefore I gave it try.

Comment (line-o) [reply exist-core/src/main/java/org/exist/xquery/value/Type.java:135]: If this is turns out to be feasible I would suggest to change NO_VALUE back to -1.

Comment (adamretter) [reply exist-core/src/main/java/org/exist/xquery/value/Type.java:135]: Sure. It should not matter if it is positive or negative as long as it is a valid java int.

File: exist-core/src/main/java/org/exist/xquery/value/Type.java [hunk 4]
```diff
@@ -105,95 +105,136 @@
     public final static int MAP = 102;
     public final static int ARRAY = 103;
     private final static Logger LOG = LogManager.getLogger(Type.class);
+
+    private static int NO_SUCH_VALUE = -99;
```
Comment (line-o) [inline exist-core/src/main/java/org/exist/xquery/value/Type.java:109]: If Type.NODE  value is changed to a positive value this should be set to -1 again.

File: exist-core/src/main/java/org/exist/xquery/value/Type.java (hunk 3 above)
Comment (line-o) [reply exist-core/src/main/java/org/exist/xquery/value/Type.java:135]: Well with it being negative nothing works... the test suites cannot even be started. With it being zero an exception is thrown that zero is not allowed. When set to a positive value the DB runs and can be used normally and core tests pass.

**Comment by line-o** [comment]
#3390 is not fixed by this PR

**Comment by joewiz** [comment]
@line-o Not sure I follow. Are you’re saying someone claims this PR fixed that issue?

**Comment by line-o** [comment]
@joewiz you were asked by @adamretter to test it with this PR applied. I did.

File: exist-core/src/main/java/org/exist/xquery/value/DoubleValue.java (hunk 2 above)
Comment (dizzzz) [reply exist-core/src/main/java/org/exist/xquery/value/DoubleValue.java:44]: just checking

File: exist-core/src/main/java/org/exist/xquery/value/Type.java (hunk 4 above)
Comment (dizzzz) [reply exist-core/src/main/java/org/exist/xquery/value/Type.java:109]: @adamretter ?

**Comment by joewiz** [comment]
@line-o Ah, I see, yes, he asked if it helped. You must not've seen [my note in Slack](https://exist-db.slack.com/archives/CG2MRUZ35/p1587915341416200) that I'd tested it (we should make better use of threads so overlapping conversations aren't so hard to follow), or the Context > eXist-db Version section of my report for #3390 stating that I had tested my report against that branch.

**Comment by line-o** [comment]
@joewiz yes, I noticed that after your question (re-reading if I missed something). But I think it is good info to have here, since it is mentioned.

**Comment by dizzzz** [comment]
so...... what's up?

**Comment by joewiz** [comment]
On today's community call @adamretter said he would like to try @line-o's suggestion.

File: exist-core/src/main/java/org/exist/xquery/value/Type.java (hunk 3 above)
Comment (adamretter) [reply exist-core/src/main/java/org/exist/xquery/value/Type.java:135]: @line-o Okay I checked out what you are mentioned.

I can indeed see the issue that you mention, it is because the Type is used in some places as an index into an array, and that can't be a negative value.

Unfortunately we cannot fix such a thing without changing the format of some `.dbx` files, as the value of the Type is encoded into key/values that are stored in the on-disk files. I haven't checked them all, but the first and most obvious is the `NativeValueIndex`. 

So if we wanted to fix that it would have to happen in a major release change and we would need to bump the file-version numbers of the relevant on-disk file formats.

I think if we get this one merged, we can open a second PR to do that for a 6.0.0 release or whatever

**Comment by adamretter** [comment]
@joewiz @dizzzz I checked our the suggestion of @line-o and it works indeed, but we can't do it without changing the on-disk file format of some of the .dbx files - best that we leave such a further change to a different PR for a major release.

**Comment by dizzzz** [comment]
does @line-o need to 'ok' his remarks, or I am I good to go?

**Comment by adamretter** [comment]
@dizzzz IMHO you are good to go
